### PR TITLE
Preparation and cleanup for Http2NetworkClient

### PR DESCRIPTION
**Pull request by zzmao**
Introduced config to enable/disable http2network client.
Created Http2NetworkClientFactory. 
Changed number of max stream for a channel to Long instead of Int.
Some preparation work for http2 client.

### CODE REVIEW

**Comment by zzmao** [comment]
@cgtz  You are optional to review this PR.
Will need you help to review next PR, http2 client core functionalities implementation.

File: ambry-api/src/main/java/com.github.ambry/config/Http2ClientConfig.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+   * Number of threads in a Netty event loop group. 0 means Netty will decide the number.
+   */
+  @Config(HTTP2_NETTY_EVENT_LOOP_GROUP_THREADS)
+  @Default("0")
+  public final int http2NettyEventLoopGroupThreads;
+
+  public Http2ClientConfig(VerifiableProperties verifiableProperties) {
+    idleConnectionTimeout = verifiableProperties.getLong(HTTP2_IDLE_CONNECTION_TIMEOUT, -1);
+    http2MinConnectionPerPort = verifiableProperties.getInt(HTTP2_MIN_CONNECTION_PER_PORT, 4);
+    http2MaxStreamsPerConnection = verifiableProperties.getLong(HTTP2_MAX_STREAMS_PER_CONNECTION, Long.MAX_VALUE);
```
Comment (justinlin-linkedin) [inline ambry-api/src/main/java/com.github.ambry/config/Http2ClientConfig.java]: Long.MAX_VALUE seems pretty scary? Is stream id reusable in http2?

File: ambry-network/src/main/java/com.github.ambry.network/http2/Http2NetworkClient.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
+/**
+ * Copyright 2019 LinkedIn Corp. All rights reserved.
```
Comment (justinlin-linkedin) [inline ambry-network/src/main/java/com.github.ambry.network/http2/Http2NetworkClient.java]: 2020

File: ambry-network/src/main/java/com.github.ambry.network/http2/Http2ClientMetrics.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package com.github.ambry.network.http2;
+
+import com.codahale.metrics.MetricRegistry;
+
+
+public class Http2ClientMetrics {
```
Comment (justinlin-linkedin) [inline ambry-network/src/main/java/com.github.ambry.network/http2/Http2ClientMetrics.java]: javadoc?

File: ambry-network/src/main/java/com.github.ambry.network/http2/Http2ChannelPoolMap.java [hunk 4]
```diff
@@ -28,9 +29,13 @@
   private final EventLoopGroup eventLoopGroup;
   private final SSLFactory sslFactory;
 
-  public Http2ChannelPoolMap(SSLFactory sslFactory, EventLoopGroup eventLoopGroup) {
+  private final Http2ClientConfig http2ClientConfig;
+
+  public Http2ChannelPoolMap(SSLFactory sslFactory, EventLoopGroup eventLoopGroup,
```
Comment (jsjtzyy) [inline ambry-network/src/main/java/com.github.ambry.network/http2/Http2ChannelPoolMap.java]: Could you add some java doc for this class?

File: ambry-network/src/main/java/com.github.ambry.network/http2/Http2NetworkClient.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import io.netty.channel.pool.ChannelPoolMap;
+import java.net.InetSocketAddress;
+import java.util.List;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A HTTP/2 implementation of {@link NetworkClient}. More implementation details will be added in next PR.
```
Comment (jsjtzyy) [inline ambry-network/src/main/java/com.github.ambry.network/http2/Http2NetworkClient.java]: minor: Add `TODO` before `More implementation details will be added in next PR.`

File: ambry-network/src/main/java/com.github.ambry.network/http2/Http2NetworkClient.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+ */
+public class Http2NetworkClient implements NetworkClient {
+  private static final Logger logger = LoggerFactory.getLogger(Http2NetworkClient.class);
+  private EventLoopGroup eventLoopGroup;
```
Comment (jsjtzyy) [inline ambry-network/src/main/java/com.github.ambry.network/http2/Http2NetworkClient.java]: can this be `final` ?

File: ambry-rest/src/main/java/com.github.ambry.rest/RestServerMain.java [hunk 7]
```diff
@@ -79,6 +80,9 @@ public static void main(String[] args) {
    * @throws IOException
    */
   private static SSLFactory getSSLFactoryIfRequired(VerifiableProperties verifiableProperties) throws Exception {
+    if (new ClusterMapConfig(verifiableProperties).clusterMapHttp2Enable) {
```
Comment (jsjtzyy) [inline ambry-rest/src/main/java/com.github.ambry.rest/RestServerMain.java]: Can we just create one instance of `ClusterMapConfig` ? No need to create ClusterMapConfig twice in some cases.

File: ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterFactory.java [hunk 8]
```diff
@@ -86,11 +91,18 @@ public NonBlockingRouterFactory(VerifiableProperties verifiableProperties, Clust
...
     networkConfig = new NetworkConfig(verifiableProperties);
+    http2ClientConfig = new Http2ClientConfig(verifiableProperties);
     networkMetrics = new NetworkMetrics(registry);
+    http2ClientMetrics = new Http2ClientMetrics(registry);
+
     time = SystemTime.getInstance();
-    networkClientFactory = new SocketNetworkClientFactory(networkMetrics, networkConfig, sslFactory,
-        routerConfig.routerScalingUnitMaxConnectionsPerPortPlainText,
-        routerConfig.routerScalingUnitMaxConnectionsPerPortSsl, routerConfig.routerConnectionCheckoutTimeoutMs, time);
+    if (new ClusterMapConfig(verifiableProperties).clusterMapHttp2Enable) {
```
Comment (jsjtzyy) [inline ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterFactory.java]: After reviewing the whole PR, I actually wonder if we can put the `Http2ClientEnabled` config into `NetworkConfig`.   Any specific reason to keep it in the `ClusterMapConfig` ?

Comment (zzmao) [reply ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterFactory.java]: I initially tried to add this in networkConfig, the problem is DataNode:getPortConnectTo(). It doesn't have NetworkConfig context.

Another benefit is we aimed to use this to override sslEnabledDc config in ClusterMapConfig.

File: ambry-api/src/main/java/com.github.ambry/config/Http2ClientConfig.java (hunk 1 above)
Comment (zzmao) [reply ambry-api/src/main/java/com.github.ambry/config/Http2ClientConfig.java]: This config is for max concurrency streams per connection, not stream id. There is no reason to limit the concurrency. 
After I double checking http2 spec, stream id is between 1-Int.max, changing this to int and reanming it to http2MaxConcurrentStreamsPerConnection.

**Comment by zzmao** [comment]
@justinlin-linkedin @jsjtzyy Addressed your comments, please take a second look at.

**Pull request merged**